src/item_pocket.h

```diff
@@ -0,0 +1,174 @@
+
+        units::volume item_size_modifier() const;
+        units::mass item_weight_modifier() const;
+
+        item *magazine_current();
+        void casings_handle( const std::function<bool( item & )> &func );
+        bool use_amount( const itype_id &it, int &quantity, std::list<item> &used );
+        bool will_explode_in_a_fire() const;
+        bool detonate( const tripoint &p, std::vector<item> &drops );
+        bool process( const itype &type, player *carrier, const tripoint &pos, bool activate,
+                      float insulation, temperature_flag flag );
+        bool legacy_unload( player *guy, bool &changed );
+        void remove_all_ammo( Character &guy );
+        void remove_all_mods( Character &guy );
+
+        // removes and returns the item from the pocket.
+        cata::optional<item> remove_item( const item &it );
+        cata::optional<item> remove_item( const item_location &it );
+        bool spill_contents( const tripoint &pos );
+        void clear_items();
+        bool has_item( const item &it ) const;
+        item *get_item_with( const std::function<bool( const item & )> &filter );
+        void remove_items_if( const std::function<bool( item & )> &filter );
+        void has_rotten_away( const tripoint &pnt );
+
+        // tries to put an item in the pocket. returns false if failure
+        ret_val<contain_code> insert_item( const item &it );
+        void add( const item &it );
+
+        // only available to help with migration from previous usage of std::list<item>
+        std::list<item> &edit_contents();
+
+        // cost of getting an item from this pocket
+        // @TODO: make move cost vary based on other contained items
+        int obtain_cost( const item &it ) const;
+        // @relates visitable
+        VisitResponse visit_contents( const std::function<VisitResponse( item *, item * )> &func,
+                                      item *parent = nullptr );
+
+        void general_info( std::vector<iteminfo> &info, int pocket_number, bool disp_pocket_number ) const;
+        void contents_info( std::vector<iteminfo> &info, int pocket_number, bool disp_pocket_number ) const;
+
+        void load( const JsonObject &jo );
+        void serialize( JsonOut &json ) const;
+        void deserialize( JsonIn &jsin );
+
+        bool operator==( const item_pocket &rhs ) const;
+
+        bool was_loaded;
+    private:
+        pocket_type type = CONTAINER;
+        // max volume of stuff the pocket can hold
+        units::volume max_contains_volume = 0_ml;
+        // min volume of item that can be contained, otherwise it spills
+        units::volume min_item_volume = 0_ml;
+        // max weight of stuff the pocket can hold
+        units::mass max_contains_weight = 0_gram;
+        // multiplier for spoilage rate of contained items
+        float spoil_multiplier = 1.0f;
+        // items' weight in this pocket are modified by this number
+        float weight_multiplier = 1.0f;
```

Comment:
Your class `item_pocket ` contains things that are going to change during the game (item instance: player will put things into the pocket). But it also contains things that will never change: `spoil_multiplier` `min_item_volume`, ...

Those should be stored in separate objects. Your current code makes it look like a combination of the item class *and* the `itype` class. Where are you even going to take the value for `max_contains_weight` from?

Reply:
so what might be a good way to go about this, if i don't want to have ids and individual json definitions for each pocket data?

Reply:
Do we really want to avoid pocket definitions? I can see a use case for defining pockets with different properties on the same item.

Reply:
You can load pocket definitions into an `itype` instance, just like we load various item slot data (e.g. data for gun item types, comestible items types, ...). See "item_factory.cpp".

Reply:
I do load pocket data into itype (it's in the second PR) and then use the itype to put all that data into item. i guess the part that i don't really understand is how to keep the contents (which aren't just a list, they're spread out among pockets) separate from the members that don't change due to player input. i suppose i could just load two of them and put one into the other? i don't want them to be precisely like itype and item, because that would mean you need a separate json definition for pockets for every different pocket you want, where right now you just define the pockets in each item.
This also can mean that i don't even want or need to deserialize the pocket... i've give it some thought